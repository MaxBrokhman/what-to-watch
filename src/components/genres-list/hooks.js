import {setGenre} from '../../reducer/actions';

export const useGenres = (movies, dispatch) => {
  const res = [`all genres`];
  for (let movie of movies) {
    if (res.indexOf(movie.genre) === -1) {
      res.push(movie.genre);
    }
  }
  const clickHandler = (genre) => (evt) => {
    evt.preventDefault();
    setGenre(genre, dispatch);
  };
  return {
    genres: res,
    clickHandler,
  };
};
