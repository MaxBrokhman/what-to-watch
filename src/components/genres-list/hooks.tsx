import {SyntheticEvent} from 'react';

import {setGenre} from '../../reducer/actions';
import {TFilm, TDispatch} from '../../reducer/types';

type TUseGenres = {
  genres: Array<string>;
  clickHandler: (genre: string) => (evt: SyntheticEvent) => void;
}

export const useGenres = (movies: Array<TFilm>, dispatch: TDispatch): TUseGenres => {
  const res = [`all genres`];
  for (const movie of movies) {
    if (res.indexOf(movie.genre) === -1) {
      res.push(movie.genre);
    }
  }
  const clickHandler = (genre: string) => (evt: SyntheticEvent): void => {
    evt.preventDefault();
    setGenre(genre, dispatch);
  };
  return {
    genres: res,
    clickHandler,
  };
};
