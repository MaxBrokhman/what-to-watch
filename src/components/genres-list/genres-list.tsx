import React from 'react';

import {useAppContext} from '../../reducer/reducer';
import {useGenres} from './hooks';

export const GenresList = (): JSX.Element => {
  const {state, dispatch} = useAppContext();
  const {genres, clickHandler} = useGenres(state.filmsList, dispatch);
  return (
    <ul className="catalog__genres-list">
      {
        genres.map((genre) => (
          <li className="catalog__genres-item" key={genre}>
            <a
              href="#"
              className="catalog__genres-link"
              onClick={clickHandler(genre)}
            >
              {genre}
            </a>
          </li>
        ))
      }
    </ul>
  );
};
