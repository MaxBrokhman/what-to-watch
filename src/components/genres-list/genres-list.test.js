import React from 'react';
import renderer from 'react-test-renderer';

import {GenresList} from './genres-list';

it(`Renders GenresList component correctly`, () => {
  const tree = renderer
                .create(<GenresList />)
                .toJSON();
  expect(tree).toMatchSnapshot();
});
