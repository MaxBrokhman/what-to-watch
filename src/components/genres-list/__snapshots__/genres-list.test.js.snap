// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders GenresList component correctly 1`] = `
<ul
  className="catalog__genres-list"
>
  <li
    className="catalog__genres-item"
  >
    <a
      className="catalog__genres-link"
      href="#"
      onClick={[Function]}
    >
      all genres
    </a>
  </li>
  <li
    className="catalog__genres-item"
  >
    <a
      className="catalog__genres-link"
      href="#"
      onClick={[Function]}
    >
      fantasy
    </a>
  </li>
  <li
    className="catalog__genres-item"
  >
    <a
      className="catalog__genres-link"
      href="#"
      onClick={[Function]}
    >
      drama
    </a>
  </li>
  <li
    className="catalog__genres-item"
  >
    <a
      className="catalog__genres-link"
      href="#"
      onClick={[Function]}
    >
      thriller
    </a>
  </li>
  <li
    className="catalog__genres-item"
  >
    <a
      className="catalog__genres-link"
      href="#"
      onClick={[Function]}
    >
      comedy
    </a>
  </li>
</ul>
`;
